kml/minzoom_quadtree.hpp

```diff
@@ -0,0 +1,152 @@
+#pragma once
+#include "kml/types.hpp"
+
+#include "indexer/scales.hpp"
+
+#include "geometry/mercator.hpp"
+#include "geometry/rect2d.hpp"
+
+#include "base/assert.hpp"
+#include "base/bits.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <iterator>
+#include <limits>
+#include <utility>
+#include <vector>
+
+namespace kml
+{
+// Builds linear quadtree of input values using coordinates provided
+// then walks through the quadtree elevating minZoom of those values
+// which are greater (in sense of Less) then others on a quadtree level.
+// To encode each quadtree level uses two bits. One from ordinate and
+// another from abscissa. Both components are mapped to a coordinate
+// system that uses the full range of uint32_t inside the bounding (square) box.
+template <typename Value, typename Less>
+class MinZoomQuadtree
+{
+public:
+  MinZoomQuadtree(Less const & less) : m_less{less} {}
+
+  template <typename ValueType>
+  void Add(m2::PointD const & point, ValueType && value) { m_quadtree.push_back({point, std::forward<ValueType>(value)}); }
+
+  void Clear() { m_quadtree.clear(); }
+
+  template <typename F>
+  void SetMinZoom(double countPerTile, int maxZoom, F setMinZoom /* void (Value & value, int minZoom) */)
+  {
+    CHECK_GREATER(countPerTile, 0.0, ());
+    CHECK_LESS_OR_EQUAL(maxZoom, scales::GetUpperStyleScale(), ());
+    CHECK_GREATER_OR_EQUAL(maxZoom, 1, ());
+
+    if (m_quadtree.empty())
+      return;
+
+    m2::RectD bbox;
+    for (auto const & elem : m_quadtree)
+      bbox.Add(elem.m_point);
+
+    if (bbox.IsEmptyInterior())
+      return;
+
+    auto spacing = std::max(bbox.SizeX() / mercator::Bounds::kRangeX,
+                            bbox.SizeY() / mercator::Bounds::kRangeY);
+    spacing *= std::sqrt(countPerTile);
+
+    int baseZoom;
+    auto const scale = std::frexp(spacing, &baseZoom);
+    baseZoom = 1 - baseZoom;
```

Comment:
Can we have a comment for these two lines, please?

Reply:
Done